espnet2/train/trainer.py

```diff
@@ -210,10 +211,14 @@ def run(
         else:
             scaler = None
 
-        use_lora = getattr(trainer_options, "use_lora", False)
-        save_lora_only = getattr(trainer_options, "save_lora_only", False)
-        if use_lora and lora is None:
-            raise RuntimeError("Requiring loralib. Do 'pip install loralib'")
+        adapter = getattr(trainer_options, "adapter", None)
+        use_adapter = getattr(trainer_options, "use_adapter", False)
+        save_adapter_only = getattr(trainer_options, "save_adapter_only", False)
+        if use_adapter:
+            if adapter == "lora" and lora is None:
+                raise RuntimeError("Requiring loralib. Do 'pip install loralib'")
+
+            # TODO: houlsby adapter may need S3PRL?
```

Comment:
Is it possible to make it general to other modules as well? If that's difficult, we may keep only to s3prl for now.

In that case, you may consider add a check to ensure that s3prl is installed when using the houlsby adapter.

Reply:
Is this update compatible with pretrained models which have configs `use_lora` and `save_lora_only`?

Reply:
Sorry for the late reply.
@simpleoier I think it's compatible but need to do some revision.
And I did it by rename `use_lora` to `use_adapter` (more general), and the same way with `save_lora_only`.
The reason is that in most case, we add lora only on the pre-trained model. However, in SSL settings, most of the time we need to reinitialize a downstream model for each task. That is to say, when applying lora to SSL models as adapters, we need to save not only the [lora parameters](https://github.com/espnet/espnet/blob/b7b37f5689fa4c065f2b1a0df3f9b7da12190093/espnet2/train/trainer.py#L365) but also the downstream model and other tunable parameters. Therefore, I choose to save all parameters `requires_grad = True`.

Reply:
I made some revisions here.
Please check [d7f0b39](https://github.com/espnet/espnet/pull/5609/commits/d7f0b399dd47ccb7dd2714b0b584fad8b5a3c2be)